emscripten/native/ui/MapView.cpp

```diff
@@ -0,0 +1,239 @@
+#include "ui/MapView.h"
+#include "ui/BaseMapView.h"
+#include "ui/MapEventListener.h"
+#include "ui/MapLicenseManagerListener.h"
+#include "ui/MapRedrawRequestListener.h"
+#include "ui/EmscriptenInput.h"
+#include "components/Options.h"
+#include "components/Layers.h"
+#include "core/MapBounds.h"
+#include "core/MapPos.h"
+#include "core/MapVec.h"
+#include "core/ScreenPos.h"
+#include "core/ScreenBounds.h"
+#include "utils/Const.h"
+#include "renderers/MapRenderer.h"
+
+#include <emscripten.h>
+#include <emscripten/html5.h>
+
+#include <cmath>
+
+namespace carto {
+    std::vector<MapView*> cartoEmscriptenMapViews;
+    bool cartoEmscriptenMainLoopStarted = false;
+    bool waitResizeCount = 0;
+
+    void _emscripten_main_loop() {
+        for (MapView* mapView : cartoEmscriptenMapViews) {
+            if (mapView->_needRedraw) {
+                mapView->_needRedraw = false;
+                mapView->onDrawFrame();
+            }
+        }
+        if (waitResizeCount != 0) waitResizeCount -= 1;
+    }
+
+    EM_BOOL _emscripten_resize_callback(int eventType, const EmscriptenUiEvent *uiEvent, void *userData) {
+        if (waitResizeCount == 0) {
+            waitResizeCount = 10; // wait 10 frames
+            for (MapView* mapView : cartoEmscriptenMapViews) {
+                mapView->onSurfaceChanged();
+            }
+        } 
+        
+        return true;
+    }
+
+    MapView::MapView(std::string canvasId) {
+        _canvasId = canvasId;
+        _baseMapView = std::make_shared<BaseMapView>();
+        // _scale = emscripten_get_device_pixel_ratio();
+        // _scale = 2;
+
+        _baseMapView->getOptions()->setDPI(Const::UNSCALED_DPI * _scale);
+        // _baseMapView->getOptions()->setTileThreadPoolSize(2);
+        // _baseMapView->getOptions()->setEnvelopeThreadPoolSize(8);
+        // printf("dpi: %lf\n", Const::UNSCALED_DPI * _scale);
+
+        _redrawRequestListener = std::make_shared<MapRedrawRequestListener>(this);
+        _baseMapView->setRedrawRequestListener(_redrawRequestListener);
+    }
+
+    MapView::~MapView() {
+        cartoEmscriptenMapViews.erase(
+            std::remove(cartoEmscriptenMapViews.begin(), cartoEmscriptenMapViews.end(), this),
+            cartoEmscriptenMapViews.end()
+        ); 
+    }
+
+    bool MapView::registerLicense(std::string licenseKey) {
+        auto listener = std::make_shared<MapLicenseManagerListener>();
+        return BaseMapView::RegisterLicense(licenseKey, listener);
+    }
+
+    void MapView::onSurfaceCreated() {
+        // emscripten_set_canvas_element_size("#canvas", 700, 700);
+        EmscriptenWebGLContextAttributes attr;
+        emscripten_webgl_init_context_attributes(&attr);
+        attr.alpha = attr.depth = attr.stencil = attr.antialias = 
```

Comment:
This seems wrong. SDK uses alpha channel, depth buffer and stencil buffer. Thus `attr.alpha = attr.depth = attr.stencil = 1` should be used here.

Reply:
You are right. This is what causes the 3d model transparency issue. I changed it.

Reply:
Stencil creates a problem in Chrome Mobile, so I made it optional to turn it on and off. It can be given as a parameter to the MapView constructor.